### PR TITLE
Fix/280 fix clean up the spaghetti in eventsviewspy

### LINKED ISSUE #280: fix: Clean up the spaghetti in events/views.py
Big pasta in there!

We need to have a proper Tagliatelle night to fix it.

### PR DESCRIPTION
**Pull request by Theoabw**
Check `avec_data = form.cleaned_data.copy()` and if it needs deepcopy or not. The functionality should basically be the same otherwise.

Closes #280

### CODE REVIEW

**Comment by andreedd** [comment]
Would probably be good at some point to split up the validation in some kind of validation file. A bit like what was done on polls.

**Comment by theoabw** [comment]
> he functionality should basically be the same otherwise.

I totally didn't say that. Anyway, tests have passed after wayy too many tries. My fleshed out test member in tests.py is kind of unnecessary for this pr but might be useful in the future. I've tested it on localhost and basic event signup seems to be working fine. Logs don't give any weird errors at least.

**Comment by theoabw** [comment]
I still don't quite know how

```
if self.object.sign_up and (request.user.is_authenticated
                                    and self.object.registration_is_open_members()
                                    and Member.objects.get(username=request.user.username).get_active_subscription() is not None
                                    or self.object.registration_is_open_others()
                                    or request.user.groups.filter(
                    name="commodore").exists()):
```
                    
In the original code didn't throw an exception and my refactored version that defined the bools beforehand, but kept the if statement the same, did?

File: events/views.py [hunk 1]
```diff
@@ -58,119 +58,120 @@ def get_context_data(self, **kwargs):
...
-        if redirect_link and show_content:
-            return redirect(redirect_link)
+    def get_template_names(self):
+        event_title = self.get_context_data().get('event').title.lower()
+        logger.debug(event_title)
+        if event_title in ['årsfest', 'årsfest gäster']:
+            return ['events/arsfest.html']
+        if self.object.passcode and self.object.passcode != self.request.session.get('passcode_status', False):
+            return ['events/event_passcode.html']
+        return [self.template_name]
```
Comment (jaspnas) [inline events/views.py:76]: This wasn't a list previously, is this correct?

File: events/tests.py [hunk 2]
```diff
@@ -3,12 +3,41 @@
...
+
+        subscription = Subscription.objects.create(
+            name="Basic Subscription",
+            does_expire=True,
+            renewal_scale="year",
+            renewal_period=1,
+            price=100.00
+        )
+
+        SubscriptionPayment.objects.create(
```
Comment (jaspnas) [inline events/tests.py]: Would be nice to have this as a variable for the class as well to be able to adjust the subscription status

File: events/tests.py [hunk 3]
```diff
@@ -3,12 +3,41 @@
...
+            email="testuser@example.com",
+            first_name="Test",
+            last_name="User",
+            phone="123456789",
+            address="123 Test Street",
+            zip_code="00000",
+            city="Test City",
+            country="Finland",
+            membership_type=ORDINARY_MEMBER,  # Using one of your defined membership types
+            is_superuser=True
```
Comment (jaspnas) [inline events/tests.py]: Set this to false, superusers should not sign up for events in production

**Review by jaspnas — DISMISSED** [review]
Resolve the comments and rebase, looks like it's working for most normal use cases

**Review by jaspnas — DISMISSED** [review]
Needs more tests, should at least test the different types of members with different subscription payment statuses.

**Comment by theoabw** [comment]
> Needs more tests, should at least test the different types of members with different subscription payment statuses.

> Would probably be good at some point to split up the validation in some kind of validation file. A bit like what was done on polls.

I'm working on splitting up validation into a separate validation file. I noticed that the detail template was only checking for if user was validated and not if the user had an active subscription. This leads to members without an active subscription being able to see the registration form, but not being able to register to the event if self.object.registration_is_open_members() is true. 

Members without an active subscription shouldn't be able to see the registration form for an event that's only open for members.

**Comment by theoabw** [comment]
1. TODO: Captcha test
2. TODO: avec test
3. TODO: redirect link test
4. TODO: anonymous test
5. TODO: custom field test
6. TODO: max participants test

File: events/tests.py [hunk 4]
```diff
@@ -87,11 +135,134 @@ def test_get_events_index(self):
...
-        self.assertEqual(event.get_registrations().count(), 0)
+        self.assertEqual(self.event.get_registrations().count(), 0)
+
+    def test_event_detail_view_with_and_without_passcode(self):
+        self.event.passcode = 'secret'
+        self.event.save()
+        c = Client()
+        wrong, right = {'passcode': 'wrong'}, {'passcode': 'secret'}
+        incorrect = c.post(reverse('events:detail', args=[self.event.slug]), wrong)
+        self.assertEqual(incorrect.status_code, 200)
```
Comment (jaspnas) [inline events/tests.py:152]: Should this be changed to return 401

Can be done as a later issue (create and resolve this comment)

**Review by jaspnas — APPROVED** [review]

**Pull request merged**